### ISSUE TITLE
get-a-tree should be more restricitve

### ISSUE THREAD

**Issue by uliska** [openlilylib/snippets#68]
When using `get-a-tree` or `\getatree` (for example when handling openlilylib-options) `#f` is returned when retrieving a non-existent key or even when trying to access a non-existent alist.

I was about to handle this in the option handling when I realized that it should be dealt with at the source. Unfortunately the code of `get-a-tree` is somewhat too cryptic (and undercommented) for me to try fiddling around with it..

One problem is that `assoc-get` returns `#f` for a missing key - which can't be distinguished from an existing key with the value `#f`.

I'm not completely sure what this should return, as returning the pair through `assoc` adds an additional layer of complexity when processing the result. But maybe it's necessary as the ambiguity between `#f` and `#f` is not really tolerable.

**Comment by PaulMorris** [comment]
I think in guile scheme one would want to return an empty list '() instead of false #f here. Then use null? to test whether an empty list has been returned.  
https://www.gnu.org/software/guile/manual/guile.html#List-Predicates
http://stackoverflow.com/questions/9115703/null-value-in-mit-scheme

**Comment by uliska** [comment]
## Ok I'll update my pull request. Currently I'm returning #<unspecified>

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by uliska** [comment]
Ah no. An empty list would also be a valid value so it doesn't work to distinguish here.  Unspecified seems be the right thing.

**Comment by PaulMorris** [comment]
Ok. Do you know how a return of #<unspecified> would be understood and handled by the calling function? For example, I don't see an "unspecified?" predicate in the guile manual.  This is unfamiliar territory for me...

**Comment by PaulMorris** [comment]
Huh, looks like "unspecified?" works (even though it's not listed in the guile manual).   I tried this in Frescobaldi:

```
#(display (if #f #t))
#(newline)
#(display (unspecified? (if #f #t)))
```

**Comment by uliska** [comment]
> Ok. Do you know how a return of #<unspecified> would be understood and
> handled by the calling function? For example, I don't see an
> "unspecified?" predicate in the guile manual.  This is unfamiliar
> territory for me...

For me too, I don't have a clue yet. But OTOH I need a way to return "something" that makes clear that the key is unassigned when #f and '() can be consideted valid values ...

**Comment by jpvoigt** [comment]
I will have a look at the pull request later. Returning #<unspecified>
sounds creepy, but I just saw your replies concerning "unspecified?" ...
Hi Paul and Urs, if this works, you can of course accept the pull request.
First I thought about another function 'a-list-has-key' or the like.

More on this later - family is calling ;)

Am 20.02.2015 um 15:15 schrieb Urs Liska:

**Comment by Cecca** [comment]
I agree with @jpvoigt, returning `#<unspecified>` is a little strange.
`#<unspecified>` is not a value, that is, is not something that one can instantiate like, say, `#f` or `32` or `'(1 2 3)`, you have to resort to tricks like `(if #f #t)`. I think it's more something that is used internally by scheme to signal error conditions, and the fact that the `unspecified?` function is not documented sounds like a warning to me.

I'd suggest to mimic the behaviour of the standard `assoc` and `assoc-ref` functions here. From the [documentation](https://www.gnu.org/software/guile/docs/docs-1.8/guile-ref/Retrieving-Alist-Entries.html) of guile:

> **assoc**, **assq**, **assv**
> 
> Return the first entry in alist with the given key. The return is the pair (KEY . VALUE) from alist. If there's no matching entry the return is #f. 
> 
> **assoc-ref**, **assq-ref**, **assv-ref**
> 
> Return the value from the first entry in alist with the given key, or #f if there's no such entry.
> Notice these functions have the key argument last, like other -ref functions, but this is opposite to what what assq etc above use.
> 
> When the return is #f it can be either key not found, or an entry which happens to have value #f in the cdr. Use assq etc above if you need to differentiate these cases. 

So I think that the most idiomatic way to deal with `#f` as a value is to have another function that instead of the value alone returns
1. the pair `(KEY . VALUE)` if the pair exists (hence `(key-that-exists . #f)` in our case)
2. `#f` when the key does not exist

Hence if someone is sure that `#f` is not a value that can be associated with that key he should use the function that already exists, otherwise he should use the one that returns the `(KEY . VALUE)` pair only if the key exists. Then getting the value is a matter of calling `cdr` on the pair.

To simplify usage, I second the proposal of @jpvoigt, i.e. having a function like `a-list-has-key` that can check for the existance of the key-value pair.

**Comment by jpvoigt** [comment]
@cecca this sounds reasonable.
We should do it that way
(I'm not at a Lilypond capable machine right now )

## 

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by uliska** [comment]
> @cecca this sounds reasonable.
> We should do it that way
> (I'm not at a Lilypond capable machine right now )

So do I get this right:
We retrieve the value of a given key and get the value or #f as a return 
value.

When another function (e.g. has-key) returns #f we interpret #f as the 
value #f and when has-key returns #f we interpret #f as "unassigned". I 
find this creepy too.

Hm, other languages would also consider throwing errors instead. 
Accessing an unassigned key from a Python dictionary results in a 
KeyError (if not worked around with the dict.key() function.

So what would you think? wouldn't it be even more consistent if 
requesting a non-existent key would result in no return value but a 
(ly:error) instead?

**Comment by jpvoigt** [comment]
Hm, throwing an error would not be, what I want in most cases. I would recommend the assoc solution - another function returning the key-value-pair or #f.

**Comment by flaminghakama** [comment]
On Fri, Feb 20, 2015 at 11:02 AM, Urs Liska notifications@github.com
wrote:

> So what would you think? wouldn't it be even more consistent if
> requesting a non-existent key would result in no return value but a
> (ly:error) instead?

+1

**Comment by Cecca** [comment]
> When another function (e.g. has-key) returns #f we interpret #f as the value #f and when has-key returns #f we interpret #f as "unassigned". I find this creepy too.

First of all, there are too many  `#f`s here :-) Let me rephrase: _When another function (e.g. has-key) returns #t we interpret #f as the value #f and when has-key returns #f we interpret #f as "unassigned"_.

Then, the fact is that if `has-key` returns `#f`, one shouldn't even try to interpret the value returned by `get-a-tree`. To make an example:

with the `has-key` function

``` scheme
(if (has-key tree key)
    ;; We have value, and it may possibly be (= value #f)
    (let ((value (get-a-tree tree key)))
      (do-something-with value))
    ;; We don't have value, we can't use it
    (do-something-else))
```

with the `undefined?` function

``` scheme
(let ((maybe-value (get-a-tree tree key)))
  (if (undefined? maybe-value)
      ;; We don't have value, we can't use it
      (do-something-else)
      ;; We have value, and it may possibly be (= value #f)
      (do-something-with maybe-value)))
```

So you see that the control flow is pretty much the same, only the statements are in a different order. With the `has-key` function you first check if that the key is present, and then you are sure that anything coming out of `get-a-tree` is the actual value. With the `undefined?` approach, you first get something that may or may be not the value and then you check for the actual validity.

I somehow feel a little uneasy in using an undocumented function from the core guile library. So maybe a good compromise is a solution that stands in the middle: provide a function that allows to check if a key exists and throw an error if someone tries to get a key that does not exist.

Maybe `ly:assoc-get` can come to the rescue? From the [documentation](http://lilypond.org/doc/v2.18/Documentation/source/Documentation/internals/scheme-functions)

> **ly:assoc-get** _key alist default-value strict-checking_
> 
> Return value if key in alist, else default-value (or #f if not specified). If strict-checking is set to #t and key is not in alist, a programming_error is output.

**Comment by flaminghakama** [comment]
On Fri, Feb 20, 2015 at 11:28 AM, jpvoigt notifications@github.com wrote:

> Hm, throwing an error would not be, what I want in most cases. I would
> recommend the assoc solution - another function returning the
> key-value-pair or #f.

It might make sense to always return a pair.  That way, the context of the
value is clear.

How about returning a pair with the value being an errror?
Otherwise, it seems like we'd need some special value that means "requested
key is not defined",  as the value-part of the pair.

Elaine

**Comment by uliska** [comment]
@Cecca wrote:

> > @uliska:
> > When another function (e.g. has-key) returns #f we interpret #f as the value #f and when has-key returns #f we interpret #f as "unassigned". I find this creepy too.
> 
> First of all, there are too many #fs here :-) Let me rephrase: When another function (e.g. has-key) returns #t we interpret #f as the value #f and when has-key returns #f we interpret #f as "unassigned".

Correct, sorry for the sloppy description.

> Then, the fact is that if has-key returns #f, one shouldn't even try to interpret the value returned by get-a-tree. To make an example:
> 
> with the has-key function  
> ...
> with the undefined? function
> ...
> 
> So you see that the control flow is pretty much the same, only the statements are in a different order. With the has-key function you first check if that the key is present, and then you are sure that anything coming out of get-a-tree is the actual value. With the undefined? approach, you first get something that may or may be not the value and then you check for the actual validity.

I see. However, what I don't really like is that what `has-key` has to do is practically the same as what `get-a-key` does so we'd perform that twice. And: there still is ambiguity about what `get-a-tree` evaluating to `#f` means.

---

@flaminghakama wrote:

> > On Fri, Feb 20, 2015 at 11:28 AM, jpvoigt notifications@github.com wrote:
> > Hm, throwing an error would not be, what I want in most cases. I would
> > recommend the assoc solution - another function returning the
> > key-value-pair or #f.
> 
> It might make sense to always return a pair.  That way, the context of the
> value is clear.

I think this might actually be the most consistent behaviour, but that would significantly change the way one has to use `get-a-tree` (as said somewhere I can't really judge the implication on existing code (or rather: how much code exists that uses this function)).

The function would then either return a pair `#'(key . value)` if the key exists or `#f` if not. The user would then write something like:

``` scheme
(let ((val (get-a-tree option-list option-path)))
    (if val
        (do-some-thing-with (cdr val))
        (do-something-like-raising-an-error)))
```

I think Jan-Peter should decide. I just need to get this right to continue working on the openlilylIb option handling (next step is adding support for typed options).

**Comment by PaulMorris** [comment]
I'm liking Cecca's approach of going with more idiomatic patterns and avoiding use of "unspecified?".  ly:assoc-get looks interesting.

**Comment by Cecca** [comment]
This issue has been open for quite some time now, so I decided to tackle it myself as I had some time available :-) I opened a pull request #139 with a refactoring of the tree data structure.

With this new tree we have three lookup functions:
- `(tree-get-node tree path)` that returns the node on success and `#f` on failure
- `(tree-get tree path default)` that returns the value on success, `default` on failure if specified and throws a `'key-error` otherwise
- `(tree-lookup tree path success-fn failure-fn)` that invokes `success-fn` with the value as argument on success, and invokes `failure-fn` with the path as argument on failure.

For a detailed overview of the API, please have a look at the comments on top of [tree.scm](https://github.com/openlilylib/openlilylib/blob/97d55d46d91f7130cc86d292fe1ef5b66875833a/ly/_internal/utilities/tree.scm)

**Comment by jpvoigt** [comment]
Hi Matteo,

I'll have a look at it later. And I really don't take it rude  ;) 
Thank you for your work on it!

Cheers, Jan-Peter 

Am 13. November 2015 11:02:02 MEZ, schrieb Matteo Ceccarello notifications@github.com:

> This issue has been open for quite some time now, so I decided to
> tackle it myself as I had some time available :-) I opened a pull
> request #139 with a refactoring of the tree data structure.
> 
> With this new tree we have three lookup functions:
> - `(tree-get-node tree path)` that returns the node on success and `#f`
>   on failure
> - `(tree-get tree path default)` that returns the value on success,
>   `default` on failure if specified and throws a `'key-error` otherwise
> - `(tree-lookup tree path success-fn failure-fn)` that invokes
>   `success-fn` with the value as argument on success, and invokes
>   `failure-fn` with the path as argument on failure.
> 
> For a detailed overview of the API, please have a look at the comments
> on top of
> [tree.scm](https://github.com/openlilylib/openlilylib/blob/97d55d46d91f7130cc86d292fe1ef5b66875833a/ly/_internal/utilities/tree.scm)
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/openlilylib/openlilylib/issues/68#issuecomment-156383060

## 

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by Cecca** [comment]
I edited all the files with a misconfigured Emacs, so I inadvertently used tabs instead of spaces. I fixed that , so if you want to take a look at any file, please do so on the tip of branch cecca/tree-refactor ([here](https://github.com/openlilylib/openlilylib/blob/cecca/tree-refactor/ly/_internal/utilities/tree.scm) you have the latest version of tree.scm)

**Comment by uliska** [comment]
Closed through #139
